### DISCUSSION TITLE
Native Deno front-end bundler?

### DISCUSSION THREAD

**Discussion by reggi** [denoland/deno#15503] [Polls]
I'm very interested in writing front-end browser js using `Deno` and modules. 

* `Deno` global would have to be accounted for in some way.
* With the rise of `Vite` is this something that `Deno` would want to do?
* Would `Deno` ever consider parsing `.svelte` / `.vue` / `.astro` files through imports? Perhaps provide guidelines on how to write these in `Deno` and having the module built be served in `.ts`? Are there any `Deno` package servers that can take say a `YAML` file and convert it to a `.ts` file and cache it on the server? Perhaps some structure for passing these files through as modules (`.svelte.ts`, `.vue.ts`, `.astro.ts`, even `.html.ts`)?
* Is there already a way to create a minified web-compatible js?

My concern is that with the rise of js / ts in custom files vue astro svelte that compile to typescript but aren't, it will prevent deno as viably being used to write front-end code because these modules won't really work as normal deno modules unless you convert them to ts and distribute those.

One idea is to use a server to parse raw files and convert them to typescript on the fly something like `https://deno.land/convert/svelte`

```
import Thing from "https://deno.land/convert/svelte?source=https://deno.land/x/my-svelte-module/example.svelte";
```

Perhaps with an import map you can convert `https://deno.land/x/my-svelte-module/example.svelte` to `https://deno.land/convert/svelte?source=https://deno.land/x/my-svelte-module/example.svelte`

Perhaps deno support for`vite` + `esbuild` will be enough. 

I'm just a little confused about 3rd deno module imports within non-ts files and how bundlers will work within deno modules.

Here's an example from `svelte` importing a `npm` package. 

```svelte
<script>
	import { marked } from 'marked';
	let value = `Some words are *italic*, some are **bold**`;
</script>

{@html marked(value)}

<textarea value={value}></textarea>

<style>
	textarea { width: 100%; height: 200px; }
</style>
```

If this is to a single deno file `ts` file granted that would work, the bundler can ignore URLS and pass the build to deno (like astro does). But if this is to another svelte file like `https://deno.land/x/my-svelte-module/example.svelte`? That's not really possible. But it is possible with `npm` in a package like `my-svelte-module/example.svelte`.

Deno has support for static files via `fs.readFile` and `import.meta`, but it doesn't have the ability to crawl dependencies recursively for non `.ts` files. 

https://github.com/vitejs/vite/issues/109
https://dev.to/vonheikemen/vite-and-deno-an-experiment-1jdf
https://www.youtube.com/watch?v=jwtkxSLJqlk

**Comment by kitsonk** [comment]
esbulid supports Deno: https://esbuild.github.io/getting-started/#deno

That is a modern bundler for the front-end.

**Comment by reggi** [comment]


**Reply by kitsonk** [reply]
Again, vite support is almost there and esbuild is supported. What are you suggesting outside of popular bundles supporting Deno?
